BlueDBOnDisk/src/main/java/org/bluedb/disk/collection/ReadWriteBlueCollectionOnDisk.java

```diff
@@ -0,0 +1,212 @@
+import org.bluedb.api.BlueQuery;
+import org.bluedb.api.Mapper;
+import org.bluedb.api.Updater;
+import org.bluedb.api.exceptions.BlueDbException;
+import org.bluedb.api.index.BlueIndex;
+import org.bluedb.api.index.KeyExtractor;
+import org.bluedb.api.keys.BlueKey;
+import org.bluedb.api.keys.ValueKey;
+import org.bluedb.disk.ReadWriteBlueDbOnDisk;
+import org.bluedb.disk.collection.index.ReadWriteBlueIndexOnDisk;
+import org.bluedb.disk.collection.index.IndexManager;
+import org.bluedb.disk.collection.metadata.ReadWriteCollectionMetaData;
+import org.bluedb.disk.collection.task.BatchChangeTask;
+import org.bluedb.disk.collection.task.BatchDeleteTask;
+import org.bluedb.disk.collection.task.DeleteTask;
+import org.bluedb.disk.collection.task.InsertTask;
+import org.bluedb.disk.collection.task.ReplaceTask;
+import org.bluedb.disk.collection.task.UpdateTask;
+import org.bluedb.disk.executors.BlueExecutor;
+import org.bluedb.disk.file.ReadWriteFileManager;
+import org.bluedb.disk.query.BlueQueryOnDisk;
+import org.bluedb.disk.recovery.RecoveryManager;
+import org.bluedb.disk.segment.Range;
+import org.bluedb.disk.segment.ReadWriteSegment;
+import org.bluedb.disk.segment.ReadWriteSegmentManager;
+import org.bluedb.disk.segment.SegmentSizeSetting;
+import org.bluedb.disk.segment.rollup.RollupScheduler;
+import org.bluedb.disk.segment.rollup.RollupTarget;
+import org.bluedb.disk.segment.rollup.Rollupable;
+
+public class ReadWriteBlueCollectionOnDisk<T extends Serializable> extends ReadableBlueCollectionOnDisk<T> implements BlueCollection<T>, Rollupable {
+
+	private final BlueExecutor sharedExecutor;
+	private final String collectionKey;
+	private final RollupScheduler rollupScheduler;
+	private final RecoveryManager<T> recoveryManager;
+	private final ReadWriteFileManager fileManager;
+	private final ReadWriteSegmentManager<T> segmentManager;
+	protected final IndexManager<T> indexManager;
+
+	public ReadWriteBlueCollectionOnDisk(ReadWriteBlueDbOnDisk db, String name, Class<? extends BlueKey> requestedKeyType, Class<T> valueType, List<Class<? extends Serializable>> additionalRegisteredClasses) throws BlueDbException {
+		this(db, name, requestedKeyType, valueType, additionalRegisteredClasses, null);
+	}
+
+	public ReadWriteBlueCollectionOnDisk(ReadWriteBlueDbOnDisk db, String name, Class<? extends BlueKey> requestedKeyType, Class<T> valueType, List<Class<? extends Serializable>> additionalRegisteredClasses, SegmentSizeSetting segmentSize) throws BlueDbException {
+		super(db, name, requestedKeyType, valueType, additionalRegisteredClasses, segmentSize);
+		sharedExecutor = db.getSharedExecutor();
+		collectionKey = getPath().toString();
+		rollupScheduler = new RollupScheduler(this);
+		rollupScheduler.start();
+		fileManager = new ReadWriteFileManager(serializer);
+		recoveryManager = new RecoveryManager<T>(this, getFileManager(), getSerializer());
+		Rollupable rollupable = null;
+		rollupable = (Rollupable) this;

```

Comment:
This is a bit weird how we declare the rollupable variable, set it to null, then set it to this, then pass it into the segment manager constructor. I think `Rollupable rollupable = this` should be sufficient if you don't want to pass `this` straight into the constructor. Which makes sense since passing `this` into the constructor hides the fact that you're really passing in a rollupable.

Reply:
That's pretty funny.  I'm changing it to "Rollupable rollupable = this;" in the next commit.